p2p/src/main/kotlin/maru/p2p/discovery/MaruDiscoveryService.kt

```diff
@@ -0,0 +1,183 @@
+import org.ethereum.beacon.discovery.DiscoverySystem
+import org.ethereum.beacon.discovery.DiscoverySystemBuilder
+import org.ethereum.beacon.discovery.schema.EnrField
+import org.ethereum.beacon.discovery.schema.NodeRecord
+import org.ethereum.beacon.discovery.schema.NodeRecordBuilder
+import org.ethereum.beacon.discovery.schema.NodeRecordFactory
+import org.hyperledger.besu.metrics.noop.NoOpMetricsSystem
+import tech.pegasys.teku.infrastructure.async.AsyncRunner
+import tech.pegasys.teku.infrastructure.async.Cancellable
+import tech.pegasys.teku.infrastructure.async.MetricTrackingExecutorFactory
+import tech.pegasys.teku.infrastructure.async.SafeFuture
+import tech.pegasys.teku.infrastructure.async.ScheduledExecutorAsyncRunner
+import tech.pegasys.teku.networking.p2p.discovery.discv5.SecretKeyParser
+
+class MaruDiscoveryService(
+  privateKeyBytes: ByteArray,
+  private val p2pConfig: P2P,
+  private val forkIdProvider: () -> ForkId, // returns the ForkId of the tip of the chain
+) {
+  companion object {
+    val BOOTNODE_REFRESH_DELAY: Duration = Duration.ofMinutes(2L)
+  }
+
+  private val log: Logger = LogManager.getLogger(this::javaClass)
+
+  private var discoverySystem: DiscoverySystem
+
+  private val privateKey = SecretKeyParser.fromLibP2pPrivKey(Bytes.wrap(privateKeyBytes))
+
+  private val bootnodes =
+    p2pConfig.bootnodes
+      .stream()
+      .map { NodeRecordFactory.DEFAULT.fromEnr(it) }
+      .toList()
+
+  val delayedExecutor: AsyncRunner =
+    ScheduledExecutorAsyncRunner.create(
+      "DiscoveryService",
+      1,
+      1,
+      5,
+      MetricTrackingExecutorFactory(NoOpMetricsSystem()),
+    )
+
+  private lateinit var bootnodeRefreshTask: Cancellable
+
+  init {
+    val discoveryNetworkBuilder = DiscoverySystemBuilder()
+
+    discoveryNetworkBuilder.listen(p2pConfig.ipAddress, p2pConfig.discoveryPort.toInt())
+    discoveryNetworkBuilder.secretKey(privateKey)
+    discoveryNetworkBuilder.localNodeRecord(localNodeRecord())
+    discoveryNetworkBuilder.bootnodes(bootnodes)
+
+    discoverySystem = discoveryNetworkBuilder.build()
+  }
+
+  fun start() {
+    discoverySystem
+      .start()
+      .thenRun {
+        this.bootnodeRefreshTask =
+          delayedExecutor.runWithFixedDelay(
+            { this.pingBootnodes(bootnodes) },
+            BOOTNODE_REFRESH_DELAY,
+            { error: Throwable ->
+              log.error(
+                "Failed to contact discovery bootnodes",
+                error,
+              )
+            },
+          )
+      }
+  }
+
+  fun stop() {
+    bootnodeRefreshTask.cancel()
+    discoverySystem.stop()
+  }
+
+  fun updateForkId(forkId: ForkId) { // TODO: Need to call this when the fork id changes
+    discoverySystem.updateCustomFieldValue(
+      FORK_ID_FIELD_NAME,
+      forkId.bytes,
+    )
+  }
+
+  fun searchForPeers(): SafeFuture<Collection<MaruDiscoveryPeer>> =
+    SafeFuture
+      .of(discoverySystem.searchForNewPeers())
```

Comment:
You can use `toSafeFuture()` extension from `implementation "build.linea.internal:futures"`

Reply:
done